docs/ci_parse_txt_reply.3
.\"
.\" Copyright 1998 by the Massachusetts Institute of Technology.
.\" SPDX-License-Identifier: MIT
.\"
.TH CI_PARSE_TXT_REPLY 3 "27 October 2009"
.SH NAME
ci_parse_txt_reply \- Parse a reply to a DNS query of type TXT
.SH SYNOPSIS
.nf
#include <ci.h>

int ci_parse_txt_reply(const unsigned char* \fIabuf\fP, int \fIalen\fP,
                         struct ci_txt_reply **\fItxt_out\fP);

int ci_parse_txt_reply_ext(const unsigned char* \fIabuf\fP, int \fIalen\fP,
                             struct ci_txt_ext **\fItxt_out\fP);
.fi
.SH DESCRIPTION
The \fIci_parse_txt_reply(3)\fP and \fIci_parse_txt_reply_ext(3)\fP
functions parse the response to a query of type TXT into a linked list (one
element per sub-string) of
.IR "struct ci_txt_reply" " (" "struct ci_txt_ext" ")"
The parameters
.I abuf
and
.I alen
give the contents of the response.  The result is stored in allocated
memory and a pointer to it stored into the variable pointed to by
.IR txt_out .
It is the caller's responsibility to free the resulting
.IR txt_out
structure when it is no longer needed using the function
\fBci_free_data(3)\fP.

The structure 
.I ci_txt_reply
contains the following fields:
.nf
struct ci_txt_reply {
  struct ci_txt_reply  *next;
  unsigned int  length;
  unsigned char *txt;
};
.fi

The structure
.I ci_txt_ext
contains the following fields:
.nf
struct ci_txt_ext {
  struct ci_txt_ext  *next;
  unsigned int  length;
  unsigned char *txt;
  unsigned char record_start;
};
.fi
The
.I record_start
field in
.I struct ci_txt_ext
is 1 if this structure is a start of a TXT record, and 0 if the structure is a
continuation of a previous record. The linked list of the
.I struct ci_txt_ext
will have at least one item with
.I record_start
equal to 1, and may have some items with
.I record_start
equal to 0 between them.

These sequences of
.I struct ci_txt_ext
(starting from the item with
.I record_start
equal to 1, and ending right before the record start item) may be treated as
either components of a single TXT record or as a multi-parted TXT record,
depending on particular use case.
.SH RETURN VALUES
.BR "ci_parse_txt_reply" " (" "ci_parse_txt_reply_ext" ")"
can return any of the following values:
.TP 15
.B CI_SUCCESS
The response was successfully parsed.
.TP 15
.B CI_EBADRESP
The response was malformatted.
.TP 15
.B CI_ENODATA
The response did not contain an answer to the query.
.TP 15
.B CI_ENOMEM
Memory was exhausted.
.SH AVAILABILITY
This function was first introduced in c-ci version 1.7.0.
.SH SEE ALSO
.BR ci_query (3)
.BR ci_free_data (3)


RELEASE-NOTES.md
## c-ci version 1.34.2 - October 15 2024

This release contains a fix for downstream packages detecting the c-ci
version based on the contents of the header file rather than the
distributed pkgconf or cmake files.

## c-ci version 1.34.1 - October 9 2024

This release fixes a packaging issue.


## c-ci version 1.34.0 - October 9 2024

This is a feature and bugfix release.

Features:
* adig: read arguments from adigrc.
  [PR #856](https://github.com/c-ci/c-ci/pull/856)
* Add new pending write callback optimization via `ci_set_pending_write_cb`.
  [PR #857](https://github.com/c-ci/c-ci/pull/857)
* New function `ci_process_fds()`.
  [PR #875](https://github.com/c-ci/c-ci/pull/875)
* Failed servers should be probed rather than redirecting queries which could
  cause unexpected latency.
  [PR #877](https://github.com/c-ci/c-ci/pull/877)
* adig: rework command line arguments to mimic dig from bind.
  [PR #890](https://github.com/c-ci/c-ci/pull/890)
* Add new method for overriding network functions
  `ci_set_socket_function_ex()` to properly support all new functionality.
  [PR #894](https://github.com/c-ci/c-ci/pull/894)
* Fix regression with custom socket callbacks due to DNS cookie support.
  [PR #895](https://github.com/c-ci/c-ci/pull/895)
* ci_socket: set IP_BIND_ADDRESS_NO_PORT on ci_set_local_ip* tcp sockets
  [PR #887](https://github.com/c-ci/c-ci/pull/887)
* URI parser/writer for ci_set_servers_csv()/ci_get_servers_csv().
  [PR #882](https://github.com/c-ci/c-ci/pull/882)

Changes:
* Connection handling modularization.
  [PR #857](https://github.com/c-ci/c-ci/pull/857),
  [PR #876](https://github.com/c-ci/c-ci/pull/876)
* Expose library/utility functions to tools.
  [PR #860](https://github.com/c-ci/c-ci/pull/860)
* Remove `ci__` prefix, just use `ci_` for internal functions.
  [PR #872](https://github.com/c-ci/c-ci/pull/872)


Bugfixes:
* fix: potential WIN32_LEAN_AND_MEAN redefinition.
  [PR #869](https://github.com/c-ci/c-ci/pull/869)
* Fix googletest v1.15 compatibility.
  [PR #874](https://github.com/c-ci/c-ci/pull/874)
* Fix pkgconfig thread dependencies.
  [PR #884](https://github.com/c-ci/c-ci/pull/884)


Thanks go to these friendly people for their efforts and contributions for this
release:

* Brad House (@bradh352)
* Cristian Rodríguez (@crrodriguez)
* Georg (@tacerus)
* @lifenjoiner
* Shelley Vohr (@codebytere)
* 前进，前进，进 (@leleliu008)



src/lib/legacy/ci_parse_txt_reply.c
/* MIT License
 *
 * Copyright (c) 2023 Brad House
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice (including the next
 * paragraph) shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * SPDX-License-Identifier: MIT
 */

#include "ci_private.h"
#include "ci_data.h"

static int ci_parse_txt_reply_int(const unsigned char *abuf, size_t alen,
                                    ci_bool_t ex, void **txt_out)
{
  ci_status_t        status;
  struct ci_txt_ext *txt_head = NULL;
  struct ci_txt_ext *txt_last = NULL;
  struct ci_txt_ext *txt_curr;
  ci_dns_record_t   *dnsrec = NULL;
  size_t               i;

  *txt_out = NULL;

  status = ci_dns_parse(abuf, alen, 0, &dnsrec);
  if (status != CI_SUCCESS) {
    goto done;
  }

  if (ci_dns_record_rr_cnt(dnsrec, CI_SECTION_ANSWER) == 0) {
    status = CI_ENODATA;
    goto done;
  }

  for (i = 0; i < ci_dns_record_rr_cnt(dnsrec, CI_SECTION_ANSWER); i++) {
    const ci_dns_rr_t *rr =
      ci_dns_record_rr_get(dnsrec, CI_SECTION_ANSWER, i);
    size_t j;
    size_t cnt;


    if (rr == NULL) {
      /* Shouldn't be possible */
      status = CI_EBADRESP; /* LCOV_EXCL_LINE: DefensiveCoding */
      goto done;              /* LCOV_EXCL_LINE: DefensiveCoding */
    }

    /* XXX: Why Chaos? */
    if ((ci_dns_rr_get_class(rr) != CI_CLASS_IN &&
         ci_dns_rr_get_class(rr) != CI_CLASS_CHAOS) ||
        ci_dns_rr_get_type(rr) != CI_REC_TYPE_TXT) {
      continue;
    }

    cnt = ci_dns_rr_get_abin_cnt(rr, CI_RR_TXT_DATA);

    for (j = 0; j < cnt; j++) {
      const unsigned char *ptr;
      size_t               ptr_len;

      /* Allocate storage for this TXT answer appending it to the list */
      txt_curr =
        ci_malloc_data(ex ? CI_DATATYPE_TXT_EXT : CI_DATATYPE_TXT_REPLY);
      if (txt_curr == NULL) {
        status = CI_ENOMEM; /* LCOV_EXCL_LINE: OutOfMemory */
        goto done;            /* LCOV_EXCL_LINE: OutOfMemory */
      }

      /* Link in the record */
      if (txt_last) {
        txt_last->next = txt_curr;
      } else {
        txt_head = txt_curr;
      }
      txt_last = txt_curr;

      /* Tag start on first for each TXT record */
      if (ex && j == 0) {
        txt_curr->record_start = 1;
      }

      ptr = ci_dns_rr_get_abin(rr, CI_RR_TXT_DATA, j, &ptr_len);

      txt_curr->txt = ci_malloc(ptr_len + 1);
      if (txt_curr->txt == NULL) {
        status = CI_ENOMEM; /* LCOV_EXCL_LINE: OutOfMemory */
        goto done;            /* LCOV_EXCL_LINE: OutOfMemory */
      }
      memcpy(txt_curr->txt, ptr, ptr_len);
      txt_curr->txt[ptr_len] = 0;
      txt_curr->length       = ptr_len;
    }
  }

done:
  /* clean up on error */
  if (status != CI_SUCCESS) {
    if (txt_head) {
      ci_free_data(txt_head);
    }
  } else {
    /* everything looks fine, return the data */
    *txt_out = txt_head;
  }
  ci_dns_record_destroy(dnsrec);
  return (int)status;
}

int ci_parse_txt_reply(const unsigned char *abuf, int alen,
                         struct ci_txt_reply **txt_out)
{
  if (alen < 0) {
    return CI_EBADRESP;
  }
  return ci_parse_txt_reply_int(abuf, (size_t)alen, CI_FALSE,
                                  (void **)txt_out);
}

int ci_parse_txt_reply_ext(const unsigned char *abuf, int alen,
                             struct ci_txt_ext **txt_out)
{
  if (alen < 0) {
    return CI_EBADRESP;
  }
  return ci_parse_txt_reply_int(abuf, (size_t)alen, CI_TRUE,
                                  (void **)txt_out);
}
